Email Editor: Add support for core/group block [MAILPOET-6057]

mikejolley opened the pull request:
## Description

Adds group to the list of supported blocks. Disables support for group variations because flexbox causes issues in email clients.

While doing this I adjusted how contentSize is applied so groups inherit it correcrly, and I standardised some block class naming.

@costasovo could use your feedback on this one once tests are passing.

## Code review notes

_N/A_

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6057](https://mailpoet.atlassian.net/browse/MAILPOET-6057)
[MAILPOET-6052](https://mailpoet.atlassian.net/browse/MAILPOET-6052)

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6057]: https://mailpoet.atlassian.net/browse/MAILPOET-6057?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MAILPOET-6052]: https://mailpoet.atlassian.net/browse/MAILPOET-6052?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mikejolley committed:
Disable block variations for groups—we do not support flex layout

mikejolley committed:
Apply border style when border color is defined

mikejolley committed:
Add group support w/ support for contentSize

mikejolley committed:
Remove default content size from template

mikejolley commented on mailpoet/lib/EmailEditor/Engine/Renderer/ContentRenderer/ContentRenderer.php:
```diff
@@ -87,15 +87,55 @@ private function reset() {
+    $layout = $this->settingsController->getLayout();
+    $styles .= sprintf(
+      '
+      .is-layout-constrained > *:not(.alignleft):not(.alignright):not(.alignfull):not(.alignwide) {
```
I added these rules so our defined contentSize in theme.json was applied by default. Without this, everything was full width during render, but narrow in editor.

costasovo replied:
👍 I guess this might be related also to the error we observed with the latest Gutenberg version (https://mailpoet.atlassian.net/browse/MAILPOET-6052)

mikejolley committed:
Attempt to fix mock

mikejolley committed:
get_settings mock

mikejolley commented:
@costasovo I will leave this as draft. It's done but I need some help with mocking in the tests. Will sort it out Monday.

mikejolley committed:
Add get_settings to stub for theme json class

costasovo commented on mailpoet/lib/EmailEditor/Engine/Renderer/ContentRenderer/ContentRenderer.php:
```diff
@@ -87,15 +87,55 @@ private function reset() {
+
+    $styles .= $blockSupportStyles;
+
+    // Debugging for content styles. Remember these get inlined.
+    //echo '<pre>';
+    //var_dump($styles);
+    //echo '</pre>';
```
Isn't this a forgotten debug code?

costasovo committed:
Update SettingsController unit test for moved layout settings

[MAILPOET-6057]

costasovo commented:
@mikejolley I looked into the failing tests and updated the mock in a new commit. 

I played with the block, and when I inserted it into the email, it showed "Group blocks together. Select a layout:" which is probably the initial configuration, but I can't add a block. 
<img width="727" alt="Screenshot 2024-05-12 at 20 45 54" src="https://github.com/mailpoet/mailpoet/assets/1082140/8820592c-64cc-4c6f-bf61-98176958a909">

The same in the Template mode. I found that in template mode, we have an additional bug with the inline iserter popup. Instead, after clicking on (+) template content, e.g., column block, the editor ends up blank, but no error in the console. This is not related to the PR and is very likely the same/similar error as we had with the post-locked popup.

mikejolley committed:
Update heading test

mikejolley committed:
Leave main variation active

mikejolley commented:
@costasovo thanks for fixing the test. I had to put the default variation back in place to fix the UI. Upstream we should fix this or allow the default to be set that skips initial selection. For now its working though, and will work fine for default layouts and templates where we can set layout to 'constrained'.

costasovo commented:
@mikejolley Thanks for adding the variation select. I see the branch has some conflicts. Could you please rebase it?
~~Also, for some reason, I don't see an icon in the placeholder (see image). Is that expected?
<img width="704" alt="Screenshot 2024-05-15 at 7 43 59" src="https://github.com/mailpoet/mailpoet/assets/1082140/7a4fc976-7ef0-4ef2-8ee7-d661e9245d1c">~~

I see now it is a proper icon. It just looks weird because the other two variations are missing.

mikejolley committed:
Fix compatibility with gutenberg plugin

mikejolley commented:
@costasovo I've rebased and added a workaround for MAILPOET-6052. The issue was partially resolved here, but fixed completely by using the gutenberg prefixed function when present.

costasovo reviewed (approved):
@mikejolley Thank you for fixing it! I was curious where the side-effect causing that layout styles are added happens and it is here: https://github.com/WordPress/WordPress/blob/master/wp-includes/block-supports/layout.php#L524-L538

The styles are then stored in a store via `WP_Style_Engine::store_css_rule` but in Gutenberg plugin the class is WP_Style_Engine_Gutenberg and then also the store has the suffix so the styles generated by the side-effect are stored in a different place so the WP function doesn't see them.

FYI I created a follow-up task for the upstream change https://mailpoet.atlassian.net/browse/MAILPOET-6067
